File: src/server/src/main/java/com/project/application/models/post/CommentViewResource.java
package com.project.application.models.post;

import com.project.application.models.user.UserViewResource;
import com.project.domain.comment.Comment;
import io.micronaut.serde.annotation.Serdeable;

import java.time.ZonedDateTime;
import java.util.Comparator;
import java.util.List;
import java.util.UUID;

@Serdeable
public record CommentViewResource(UUID id, UserViewResource createdBy, ZonedDateTime createdAt, String text,
                                  UUID parentId, List<CommentViewResource> replies) {
    public static CommentViewResource from(Comment comment) {
        return new CommentViewResource(comment.getId(), UserViewResource.from(comment.getUser()),
                comment.getCreatedAt(), comment.getText(), comment.getParentId(),
                comment.getReplies().stream().map(CommentViewResource::from)
                        .sorted(Comparator.comparing(c -> c.createdAt)).toList());
    }
}

File: src/server/src/main/java/com/project/application/models/post/CommentViewResource.kt
package com.project.application.models.post

import com.project.application.models.user.UserViewResource
import com.project.domain.comment.Comment
import io.micronaut.serde.annotation.Serdeable
import java.time.ZonedDateTime
import java.util.UUID

@Serdeable
data class CommentViewResource(val id: UUID, val createdBy: UserViewResource, val createdAt: ZonedDateTime, val text: String?,
                               val parentId: UUID?, val replies: List<CommentViewResource>) {
    companion object {
        fun from(comment: Comment): CommentViewResource {
            return CommentViewResource(comment.id, UserViewResource.from(comment.user!!),
                    comment.createdAt, comment.text, comment.parentId,
                    comment.replies.map { c: Comment -> from(c) }
                            .sortedBy { commentViewResource: CommentViewResource -> commentViewResource.createdAt })
        }
    }
}
